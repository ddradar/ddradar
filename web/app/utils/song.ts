import type { Series, Song } from '@ddradar/core'
import { seriesSet } from '@ddradar/core'

/** LEVEL 1-19 */
export const levels = [...Array(19).keys()].map(i => i + 1)

export const seriesNames = [...seriesSet] as Series[]

/** `0`: DDR 1st, `1`: DDR 2ndMIX, ..., `19`: Dance Dance Revolution WORLD */
export const seriesIndexes = [...Array(seriesSet.size).keys()]

export function getDisplayedBPM({
  minBPM,
  maxBPM,
}: Pick<Song, 'minBPM' | 'maxBPM'>) {
  return minBPM === maxBPM
    ? (`${minBPM}` as const)
    : (`${minBPM}-${maxBPM}` as const)
}

export function shortenSeriesName(series: string) {
  return series.replace(/^(DDR |DanceDanceRevolution )\(?([^)]+)\)?$/, '$2')
}

export const getChartColor = (difficulty: number) =>
  ['blue', 'yellow', 'red', 'green', 'purple'][difficulty]
