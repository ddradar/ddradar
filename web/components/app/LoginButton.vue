<script setup lang="ts">
import type { DropdownItem } from '#ui/types'

const { t } = useI18n()
const { login } = await useEasyAuth()
const items = computed<DropdownItem[][]>(() => [
  [
    {
      label: t('login_via', { name: 'X' }),
      icon: 'i-simple-icons-x',
      click: () => login('twitter'),
    },
    {
      label: t('login_via', { name: 'LINE' }),
      icon: 'i-simple-icons-line',
      click: () => login('line'),
    },
    {
      label: t('login_via', { name: 'GitHub' }),
      icon: 'i-simple-icons-github',
      click: () => login('github'),
    },
  ],
])
</script>

<template>
  <UDropdown :items="items">
    <UButton icon="i-heroicons-arrow-right-end-on-rectangle-20-solid">
      {{ t('login') }}
    </UButton>
  </UDropdown>
</template>

<i18n lang="json">
{
  "ja": {
    "login": "ログイン",
    "login_via": "{name}でログイン"
  },
  "en": {
    "login": "Login",
    "login_via": "Login via {name}"
  }
}
</i18n>
