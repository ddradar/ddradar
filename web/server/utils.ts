import type { QueryObject } from 'ufo'

/**
 * Get string value from query object.
 * @param query Query object (returned by `useQuery(event)`)
 * @param key Query key
 * @returns Query value (if multiple, returns last value)
 */
export function getQueryString(query: QueryObject, key: string) {
  const value = query[key]
  return Array.isArray(value) ? value[value.length - 1] : value
}

/**
 * Get integer value from query object.
 * @param query Query object (returned by `useQuery(event)`)
 * @param key Query key
 * @returns Integer value or NaN
 */
export function getQueryInteger(query: QueryObject, key: string) {
  const value = getQueryString(query, key)
  if (value === undefined) return NaN
  const num = parseFloat(value)
  return Number.isInteger(num) ? num : NaN
}
