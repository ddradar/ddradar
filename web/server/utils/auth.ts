import type { UserSchema } from '@ddradar/core'
import { fetchLoginUser, fetchUser } from '@ddradar/db'
import type { H3Event } from 'h3'

import { paramsSchema } from '~/schemas/user'

export async function getLoginUserInfo(
  event: H3Event
): Promise<UserSchema | null> {
  const clientPrincipal = getClientPrincipal(event)
  if (!clientPrincipal) return null
  return await fetchLoginUser(clientPrincipal.userId)
}

/**
 * Check user visibility
 * @param event HTTP Event (needs [id] context)
 * @returns UserSchema if user is public or same as login user. otherwise null.
 */
export async function tryFetchUser(event: H3Event): Promise<UserSchema | null> {
  const result = await getValidatedRouterParams(event, paramsSchema.safeParse)
  if (!result.success) return null

  const user = await fetchUser(result.data.id)
  if (!user) return null

  const loginId = getClientPrincipal(event)?.userId ?? ''
  return user.isPublic || user.loginId === loginId ? user : null
}
