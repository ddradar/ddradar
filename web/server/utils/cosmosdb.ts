import { CosmosClient } from '@azure/cosmos'
import type { H3Event } from 'h3'

let _client: CosmosClient | undefined

export function getCosmosClient(event: H3Event): CosmosClient {
  return (_client ??= new CosmosClient(useRuntimeConfig(event).cosmosDBConn))
}
