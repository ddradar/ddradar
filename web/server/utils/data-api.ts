import type { H3Event } from 'h3'

export const $graphql = async <T>(
  event: H3Event,
  query: string,
  variables?: object,
  role?: string
) => {
  const { data } = await $fetch<{ data: T }>('/data-api/graphql', {
    method: 'POST',
    body: { query, variables },
    headers: {
      ...getRequestHeaders(event),
      ...(role ? { 'X-MS-API-ROLE': role } : {}),
    } as Record<string, string>,
  })
  return data
}

type QueryListResult<T> = {
  [key: string]:
    | {
        items: T[]
        hasNextPage: true
        endCursor: string
      }
    | {
        items: T[]
        hasNextPage: false
        endCursor: null
      }
}

export const $graphqlList = async <T>(
  event: H3Event,
  query: string,
  schema: string,
  variables?: object,
  role?: string
) => {
  const result: T[] = []
  let queryResult: QueryListResult<T>[string] = {
    items: [],
    endCursor: null,
    hasNextPage: false,
  }
  do {
    queryResult = (
      await $graphql<QueryListResult<T>>(
        event,
        query,
        {
          ...variables,
          cursor: queryResult?.hasNextPage ? queryResult.endCursor : null,
        },
        role
      )
    )[schema]
    result.push(...queryResult.items)
  } while (queryResult.hasNextPage)

  return result
}
