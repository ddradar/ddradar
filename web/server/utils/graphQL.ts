import { $fetch } from 'ofetch'

/**
 * Call Azure Static Web Apps built-in GraphQL Query requests.
 * https://learn.microsoft.com/azure/static-web-apps/database-overview
 * @param query GraphQL query
 * @param variables GraphQL variables
 * @returns Result data
 */
export function callGraphQL<T>(query: string, variables: object) {
  const headers = useRequestHeaders() as Record<string, string>
  return $fetch<{ data: T }>(`/data-api/graphql`, {
    headers,
    method: 'POST',
    body: { query, variables },
  })
}
