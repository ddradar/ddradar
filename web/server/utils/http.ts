import { createError, sendError } from 'h3'

/**
 * Returns null and send H3Error.
 * @param event HTTP Event
 * @param statusCode HTTP Status code
 * @param message Error message (optional)
 */
export function sendNullWithError(
  event: Parameters<typeof sendError>[0],
  statusCode: number,
  message?: string
) {
  sendError(event, createError({ statusCode, message }))
  return null
}
