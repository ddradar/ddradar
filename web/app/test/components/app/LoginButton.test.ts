import { mockNuxtImport, mountSuspended } from '@nuxt/test-utils/runtime'
import { RouterLinkStub } from '@vue/test-utils'
import { describe, expect, test, vi } from 'vitest'
import { createI18n } from 'vue-i18n'

import LoginButton from '~/components/app/LoginButton.vue'
import { global, locales } from '~/test/utils'

const { navigateToMock } = vi.hoisted(() => ({ navigateToMock: vi.fn() }))
mockNuxtImport('navigateTo', () => navigateToMock)

describe('components/app/LoginButton.vue', () => {
  test.each(locales)('{ locale: "%s" } snapshot test', async locale => {
    // Arrange
    const stubs = { RouterLink: RouterLinkStub }
    const global = { stubs, plugins: [createI18n({ locale, legacy: false })] }
    const wrapper = await mountSuspended(LoginButton, { global })
    // Act
    await wrapper.find('button').trigger('click') // Expand dropdown
    // Assert
    expect(wrapper.find('div').element).toMatchSnapshot()
  })

  test.each([
    [0, 'twitter'],
    [1, 'line'],
    [2, 'github'],
  ])(
    'click menu:%i calls navigateTo("/.auth/login/%s")',
    async (i, provider) => {
      // Arrange
      vi.mocked(navigateTo).mockClear()
      const wrapper = await mountSuspended(LoginButton, { global })
      // Act
      await wrapper.find('button').trigger('click') // Expand dropdown
      await wrapper.findAll("button[role='menuitem']")[i]?.trigger('click') // Click menu item
      // Assert
      expect(vi.mocked(navigateTo)).toHaveBeenCalledWith(
        `/.auth/login/${provider}`,
        {
          external: true,
        }
      )
    }
  )
})
