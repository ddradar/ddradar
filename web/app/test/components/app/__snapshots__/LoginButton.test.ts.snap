// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/app/LoginButton.vue > { locale: "en" } snapshot test 1`] = `
<div
  class="relative inline-flex text-left rtl:text-right"
  data-headlessui-state="open"
>
  <div
    aria-controls="headlessui-menu-items-nReCNjHcrOY_1"
    aria-expanded="true"
    aria-haspopup="menu"
    class="inline-flex w-full"
    data-headlessui-state="open"
    disabled="false"
    id="headlessui-menu-button-nReCNjHcrOY_0"
    role="button"
  >
    
    <button
      class="focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm gap-x-1.5 px-2.5 py-1.5 shadow-sm text-white dark:text-gray-900 bg-primary-500 hover:bg-primary-600 disabled:bg-primary-500 dark:bg-primary-400 dark:hover:bg-primary-500 dark:disabled:bg-primary-400 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-primary-500 dark:focus-visible:outline-primary-400 inline-flex items-center"
      type="button"
    >
      
      
      <svg
        aria-hidden="true"
        class="icon flex-shrink-0 h-5 w-5"
        data-v-2205d584=""
        height="1em"
        role="img"
        viewBox="0 0 20 20"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
        xmlns:xlink="http://www.w3.org/1999/xlink"
      >
        <g
          clip-rule="evenodd"
          fill="currentColor"
          fill-rule="evenodd"
        >
          <path
            d="M17 4.25A2.25 2.25 0 0 0 14.75 2h-5.5A2.25 2.25 0 0 0 7 4.25v2a.75.75 0 0 0 1.5 0v-2a.75.75 0 0 1 .75-.75h5.5a.75.75 0 0 1 .75.75v11.5a.75.75 0 0 1-.75.75h-5.5a.75.75 0 0 1-.75-.75v-2a.75.75 0 0 0-1.5 0v2A2.25 2.25 0 0 0 9.25 18h5.5A2.25 2.25 0 0 0 17 15.75z"
          />
          <path
            d="M1 10a.75.75 0 0 1 .75-.75h9.546l-1.048-.943a.75.75 0 1 1 1.004-1.114l2.5 2.25a.75.75 0 0 1 0 1.114l-2.5 2.25a.75.75 0 1 1-1.004-1.114l1.048-.943H1.75A.75.75 0 0 1 1 10"
          />
        </g>
      </svg>
      
      
      Login
      
      
      <!--v-if-->
      
      
    </button>
    
  </div>
  <div
    class="z-20 group w-48"
    data-popper-placement="bottom-end"
    style="position: fixed; left: auto; top: 0px; margin: 0px; right: 0px; bottom: auto; transform: translate(0px, 8px);"
  >
    <transition-stub
      appear="true"
      css="true"
      enteractiveclass="transition duration-100 ease-out"
      enterfromclass="transform scale-95 opacity-0"
      entertoclass="transform scale-100 opacity-100"
      leaveactiveclass="transition duration-75 ease-in"
      leavefromclass="transform scale-100 opacity-100"
      leavetoclass="transform scale-95 opacity-0"
      persisted="false"
    >
      <div>
        <!--v-if-->
        <div
          aria-labelledby="headlessui-menu-button-nReCNjHcrOY_0"
          class="relative focus:outline-none overflow-y-auto scroll-py-1 divide-y divide-gray-200 dark:divide-gray-700 ring-1 ring-gray-200 dark:ring-gray-700 rounded-md shadow-lg bg-white dark:bg-gray-800"
          data-headlessui-state="open"
          id="headlessui-menu-items-nReCNjHcrOY_1"
          role="menu"
          tabindex="0"
        >
          
          <div
            class="p-1"
            role="none"
          >
            
            
            <button
              class="group flex items-center gap-1.5 w-full px-1.5 py-1.5 text-sm rounded-md text-gray-700 dark:text-gray-200"
              data-headlessui-state=""
              href=""
              id="headlessui-menu-item-nReCNjHcrOY_2"
              role="menuitem"
            >
              
              <!---->
              <span
                class="truncate"
              >
                Login via X
              </span>
              <!--v-if-->
              
            </button>
            
            
            <button
              class="group flex items-center gap-1.5 w-full px-1.5 py-1.5 text-sm rounded-md text-gray-700 dark:text-gray-200"
              data-headlessui-state=""
              href=""
              id="headlessui-menu-item-nReCNjHcrOY_3"
              role="menuitem"
            >
              
              <!---->
              <span
                class="truncate"
              >
                Login via LINE
              </span>
              <!--v-if-->
              
            </button>
            
            
            <button
              class="group flex items-center gap-1.5 w-full px-1.5 py-1.5 text-sm rounded-md text-gray-700 dark:text-gray-200"
              data-headlessui-state=""
              href=""
              id="headlessui-menu-item-nReCNjHcrOY_4"
              role="menuitem"
            >
              
              <!---->
              <span
                class="truncate"
              >
                Login via GitHub
              </span>
              <!--v-if-->
              
            </button>
            
            
          </div>
          
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`components/app/LoginButton.vue > { locale: "ja" } snapshot test 1`] = `
<div
  class="relative inline-flex text-left rtl:text-right"
  data-headlessui-state="open"
>
  <div
    aria-controls="headlessui-menu-items-nReCNjHcrOY_6"
    aria-expanded="true"
    aria-haspopup="menu"
    class="inline-flex w-full"
    data-headlessui-state="open"
    disabled="false"
    id="headlessui-menu-button-nReCNjHcrOY_5"
    role="button"
  >
    
    <button
      class="focus:outline-none disabled:cursor-not-allowed disabled:opacity-75 flex-shrink-0 font-medium rounded-md text-sm gap-x-1.5 px-2.5 py-1.5 shadow-sm text-white dark:text-gray-900 bg-primary-500 hover:bg-primary-600 disabled:bg-primary-500 dark:bg-primary-400 dark:hover:bg-primary-500 dark:disabled:bg-primary-400 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-primary-500 dark:focus-visible:outline-primary-400 inline-flex items-center"
      type="button"
    >
      
      
      <svg
        aria-hidden="true"
        class="icon flex-shrink-0 h-5 w-5"
        data-v-2205d584=""
        height="1em"
        role="img"
        viewBox="0 0 20 20"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
        xmlns:xlink="http://www.w3.org/1999/xlink"
      >
        <g
          clip-rule="evenodd"
          fill="currentColor"
          fill-rule="evenodd"
        >
          <path
            d="M17 4.25A2.25 2.25 0 0 0 14.75 2h-5.5A2.25 2.25 0 0 0 7 4.25v2a.75.75 0 0 0 1.5 0v-2a.75.75 0 0 1 .75-.75h5.5a.75.75 0 0 1 .75.75v11.5a.75.75 0 0 1-.75.75h-5.5a.75.75 0 0 1-.75-.75v-2a.75.75 0 0 0-1.5 0v2A2.25 2.25 0 0 0 9.25 18h5.5A2.25 2.25 0 0 0 17 15.75z"
          />
          <path
            d="M1 10a.75.75 0 0 1 .75-.75h9.546l-1.048-.943a.75.75 0 1 1 1.004-1.114l2.5 2.25a.75.75 0 0 1 0 1.114l-2.5 2.25a.75.75 0 1 1-1.004-1.114l1.048-.943H1.75A.75.75 0 0 1 1 10"
          />
        </g>
      </svg>
      
      
      ログイン
      
      
      <!--v-if-->
      
      
    </button>
    
  </div>
  <div
    class="z-20 group w-48"
    data-popper-placement="bottom-end"
    style="position: fixed; left: auto; top: 0px; margin: 0px; right: 0px; bottom: auto; transform: translate(0px, 8px);"
  >
    <transition-stub
      appear="true"
      css="true"
      enteractiveclass="transition duration-100 ease-out"
      enterfromclass="transform scale-95 opacity-0"
      entertoclass="transform scale-100 opacity-100"
      leaveactiveclass="transition duration-75 ease-in"
      leavefromclass="transform scale-100 opacity-100"
      leavetoclass="transform scale-95 opacity-0"
      persisted="false"
    >
      <div>
        <!--v-if-->
        <div
          aria-labelledby="headlessui-menu-button-nReCNjHcrOY_5"
          class="relative focus:outline-none overflow-y-auto scroll-py-1 divide-y divide-gray-200 dark:divide-gray-700 ring-1 ring-gray-200 dark:ring-gray-700 rounded-md shadow-lg bg-white dark:bg-gray-800"
          data-headlessui-state="open"
          id="headlessui-menu-items-nReCNjHcrOY_6"
          role="menu"
          tabindex="0"
        >
          
          <div
            class="p-1"
            role="none"
          >
            
            
            <button
              class="group flex items-center gap-1.5 w-full px-1.5 py-1.5 text-sm rounded-md text-gray-700 dark:text-gray-200"
              data-headlessui-state=""
              href=""
              id="headlessui-menu-item-nReCNjHcrOY_7"
              role="menuitem"
            >
              
              <!---->
              <span
                class="truncate"
              >
                Xでログイン
              </span>
              <!--v-if-->
              
            </button>
            
            
            <button
              class="group flex items-center gap-1.5 w-full px-1.5 py-1.5 text-sm rounded-md text-gray-700 dark:text-gray-200"
              data-headlessui-state=""
              href=""
              id="headlessui-menu-item-nReCNjHcrOY_8"
              role="menuitem"
            >
              
              <!---->
              <span
                class="truncate"
              >
                LINEでログイン
              </span>
              <!--v-if-->
              
            </button>
            
            
            <button
              class="group flex items-center gap-1.5 w-full px-1.5 py-1.5 text-sm rounded-md text-gray-700 dark:text-gray-200"
              data-headlessui-state=""
              href=""
              id="headlessui-menu-item-nReCNjHcrOY_9"
              role="menuitem"
            >
              
              <!---->
              <span
                class="truncate"
              >
                GitHubでログイン
              </span>
              <!--v-if-->
              
            </button>
            
            
          </div>
          
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;
